### DISCUSSION TITLE
What is your perfect feature you want this plugin to have?

### DISCUSSION THREAD

**Discussion by bandantonio** [bandantonio/obsidian-apple-books-highlights-plugin#33] [Idea Hub]
If you could think of one feature that would make the experience with the plugin even more enjoyable, what would this one feature be?

**Comment by thorbjaelk** [comment]
First of all, thank you @bandantonio for a great plugin! I would love it even more if there was a feature that allowed me to add additional content to the note manually and have these additions persist across imports. As an example, I would love to be able to add sections for my own summary and key concepts of a book and maintain (and update) the imported highlights at the end of the note as indicated by the template sketch below:

```
# {{{bookTitle}}}
Author: {{{bookAuthor}}}
[Read on Apple Books](ibooks://assetid/{{bookId}})

## Summary

## Key Concepts

## Highlights
{{#each annotations}}
> {{{highlight}}}{{#if highlightLocation}}[⤴️](ibooks://assetid/{{../bookId}}#{{highlightLocation}}){{else}}{{/if}}
{{/each}}
```

The ##Summary and ##Key Concepts sections are where I would like to manually add content that would persist across imports. Currently, and please correct me if I'm wrong, I believe this content would disappear every time I run the plugin and would force me to manually copy/paste this content to keep it.

**Comment by bandantonio** [comment]
@thorbjaelk Thank you for the request.

One of the approaches I can think of that can help you implement what you want is to create a folder somewhere (outside the highlights folder) in your vault where you will save your reflections about books and then embed those reflections into the template for importing highlights.

For example, let's say your folder with reflections is called `my-book-reflections`. In this folder, you will store reflections in some pre-defined format. For example, `Key notes - <book title>`. Make sure that `<book title>` exactly matches the one from imported highlights.

Example of such a file: **Key notes - Atomic Habits - Tiny Changes, Remarkable Results**

```md
# Key notes - Atomic Habits - Tiny Changes, Remarkable Results

## Summary

Atomic Habits book summary

## Key Concepts

- Key concept #1
- Key concept #2
- Key concept #3
```

In this way, you can modify the template for importing highlights as follows:

```hbs
...
![[Key notes - {{{bookTitle}}}#Summary]]

![[Key notes - {{{bookTitle}}}#Key Concepts]]

## Highlights
..
```

As a result, this part in the imported book file will look as follows:

```
...
![[Key notes - Atomic Habits - Tiny Changes, Remarkable Results#Summary]]

![[Key notes - Atomic Habits - Tiny Changes, Remarkable Results#Key Concepts]]

## Highlights
```

Rendered version:

<img width="688" alt="rendered-embed" src="https://github.com/user-attachments/assets/d49662c6-6967-4599-aa1b-0c82a6e23600">

In this way, your manually added content will persist across imports (as it is stored "externally", and the links will automatically point to the corresponding "Key notes" files and grab the content from the corresponding sections. If the corresponding "Key notes" file doesn't exist, you will be able to create it in one click:

<img width="661" alt="Missing key notes" src="https://github.com/user-attachments/assets/e80a4e51-376d-4b42-a262-f84451779cc3">

Please let me know your thoughts on whether this solution suits you.

> [!warning]
> Make sure to create the "key notes" folder outside the highlights folder (for now, at least). Currently, the import logic ignores nested folders and the content inside them. I'm in the middle of refactoring the import logic according to #34 but I believe that the logic of backing up only the direct children of the highlights folder will remain the same.

**Reply by thorbjaelk** [reply]
@bandantonio Ideally, I'd like to have a solution where everything is contained in a single file. However, I agree that there are a couple of different ways to handle this if I disregard that constraint. I'll work with that for now. Please let me know if you would consider implementing something like this in the future. Again, thanks for an awesome plugin!

**Reply by bandantonio** [reply]
@thorbjaelk understood. I should mention that, as its name states, the pivotal purpose of the plugin is to provide a reliable way to _import_ the existing highlights to your Vault. The plugin can only _retrieve_ data from the database (not create or update), and its scope is limited to this data.
Your request falls outside of this scope as it requires the introduction of the persistence layer to store custom data. This will increase the complexity of the codebase and break the current philosophy. At this moment, there are plans to do it, but you never know what the future holds.

**Comment by bandantonio** [comment]
Hi @thorbjaelk
My previous response was poorly thought out, a bit harsh, and misleading. Please accept my apology for that.

I have some good news for you. Today, I did some research regarding your request and eventually managed to create a draft version of the template that allows you to define custom content that will persist across imports. I tested the code with only one block, it seems to be OK. So I believe that with some more polishing, I will be able to deliver that feature.

**Comment by thorbjaelk** [comment]
@bandantonio That's excellent news! Please let me know if I can support you with testing or in any other way.

**Comment by davidishee** [comment]
Excellent plugin! I would like an option to cleanup whitespace as some highlights have extra spaces and newlines in them. I'm using the https://github.com/zlovatt/obsidian-trim-whitespace plugin right now to process them one at a time, but that is slow. If there is a way to batch process all highlights after the file is created, that would also work. 

I generally don't highlight highly structured text, just sentences or fragments of sentences.

**Reply by bandantonio** [reply]
@davidishee Hi, and thanks for your feedback.
Basically, this functionality is already in place and does its work behind the scenes. You can check the thread in #28 for more context. If this functionality is not enough, or, if you have some specific cases, feel free to open a new bug report and let's discuss them in detail.

**Comment by magnus919** [comment]
I just started using this plugin. One thing I'd love to be able to do, since I already have another plugin that creates a note with front matter about books I'm reading, I'd love to be able to have shortcake within an existing page so I can embed the annotations for a book _within_ an existing note that applies to the book more broadly.

**Comment by bandantonio** [comment]
> I would love it even more if there was a feature that allowed me to add additional content to the note manually and have these additions persist across imports

> I'd love to be able to have shortcake within an existing page so I can embed the annotations for a book within an existing note that applies to the book more broadly

@thorbjaelk @magnus919 This was such a long-awaited feature and it's finally here - the recently released version [1.8.0](https://github.com/bandantonio/obsidian-apple-books-highlights-plugin/releases/tag/1.8.0) introduces the [Keep Me section](https://mister-gold.pro/obsidian-apple-books-highlights-plugin/guide/settings.html#template-keep-me-section). Can't wait for you to check it out.
